File: capacitor-testapp/ios/EkycPlugin.swift
//
//  EkycPlugin.swift
//  App
//
//  Created by Longcon99 on 22/10/24.
//

import Capacitor
import ICSdkEKYC

@objc(EkycPlugin)
public class EkycPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "EkycPlugin"
    public let jsName = "EkycPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startEkycFull", returnType: CAPPluginReturnPromise)
    ]

    private var call: CAPPluginCall?
    
    // Phương thức thực hiện eKYC luồng đầy đủ bao gồm: Chụp ảnh giấy tờ và chụp ảnh chân dung
    // Bước 1 - chụp ảnh giấy tờ
    // Bước 2 - chụp ảnh chân dung xa gần
    // Bước 3 - hiển thị kết quả
    @objc func startEkycFull(_ call: CAPPluginCall) {
        ICEKYCSavedData.shared().authorization = call.getString("accessToken") ?? ""
        ICEKYCSavedData.shared().tokenId = call.getString("tokenId") ?? ""
        ICEKYCSavedData.shared().tokenKey = call.getString("tokenKey") ?? ""
        self.call = call
        
        let objCamera = ICEkycCameraRouter.createModule() as! ICEkycCameraViewController
        objCamera.cameraDelegate = self

        // Giá trị này xác định phiên bản khi sử dụng Máy ảnh tại bước chụp ảnh chân dung luồng full. Mặc định là Normal ✓
        // - Normal: chụp ảnh chân dung 1 hướng
        // - ProOval: chụp ảnh chân dung xa gần
        objCamera.versionSdk = ProOval
        
        // Giá trị xác định luồng thực hiện eKYC
        // - full: thực hiện eKYC đầy đủ các bước: chụp mặt trước, chụp mặt sau và chụp ảnh chân dung
        // - ocrFront: thực hiện OCR giấy tờ một bước: chụp mặt trước
        // - ocrBack: thực hiện OCR giấy tờ một bước: chụp mặt trước
        // - ocr: thực hiện OCR giấy tờ đầy đủ các bước: chụp mặt trước, chụp mặt sau
        // - face: thực hiện so sánh khuôn mặt với mã ảnh chân dung được truyền từ bên ngoài
        objCamera.flowType = full
        
        // Giá trị này xác định kiểu giấy tờ để sử dụng:
        // - IdentityCard: Chứng minh thư nhân dân, Căn cước công dân
        // - IDCardChipBased: Căn cước công dân gắn Chip
        // - Passport: Hộ chiếu
        // - DriverLicense: Bằng lái xe
        // - MilitaryIdCard: Chứng minh thư quân đội
        objCamera.documentType = IdentityCard
        
        // Giá trị này dùng để đảm bảo mỗi yêu cầu (request) từ phía khách hàng sẽ không bị thay đổi.
        objCamera.challengeCode = "INNOVATIONCENTER"
        
        // Bật/Tắt Hiển thị màn hình hướng dẫn
        objCamera.isShowTutorial = true
        
        // Bật/Tắt chức năng So sánh ảnh trong thẻ và ảnh chân dung
        objCamera.isEnableCompare = true
        
        // Bật/Tắt chức năng kiểm tra che mặt
        objCamera.isCheckMaskedFace = true
        
        // Lựa chọn chức năng kiểm tra ảnh chân dung chụp trực tiếp (liveness face)
        // - NoneCheckFace: Không thực hiện kiểm tra ảnh chân dung chụp trực tiếp hay không
        // - IBeta: Kiểm tra ảnh chân dung chụp trực tiếp hay không iBeta (phiên bản hiện tại)
        // - Standard: Kiểm tra ảnh chân dung chụp trực tiếp hay không Standard (phiên bản mới)
        objCamera.checkLivenessFace = IBeta
        
        // Bật/Tắt chức năng kiểm tra ảnh giấy tờ chụp trực tiếp (liveness card)
        objCamera.isCheckLivenessCard = true
        
        // Lựa chọn chế độ kiểm tra ảnh giấy tờ ngay từ SDK
        // - None: Không thực hiện kiểm tra ảnh khi chụp ảnh giấy tờ
        // - Basic: Kiểm tra sau khi chụp ảnh
        // - MediumFlip: Kiểm tra ảnh hợp lệ trước khi chụp (lật giấy tờ thành công → hiển thị nút chụp)
        // - Advance: Kiểm tra ảnh hợp lệ trước khi chụp (hiển thị nút chụp)
        objCamera.validateDocumentType = Basic
        
        // Bật chức năng hiển thị nút bấm "Bỏ qua hướng dẫn" tại các màn hình hướng dẫn bằng video
        objCamera.isEnableGotIt = true
        
        // Ngôn ngữ sử dụng trong SDK
        objCamera.languageSdk = "icekyc_vi"
        
        // Bật/Tắt Hiển thị ảnh thương hiệu
        objCamera.isShowTrademark = true
        
        DispatchQueue.main.async {
            objCamera.modalPresentationStyle = .fullScreen
            objCamera.modalTransitionStyle = .coverVertical
            self.bridge?.viewController?.present(objCamera, animated: true)
        }
        
    }
    
}

extension EkycPlugin: ICEkycCameraDelegate {
    
    public func icEkycGetResult() {
        // Thông tin bóc tách OCR
        let ocrResult = ICEKYCSavedData.shared().ocrResult
        
        // Kết quả kiểm tra giấy tờ chụp trực tiếp (Liveness Card) mặt trước
        let livenessCardFrontResult = ICEKYCSavedData.shared().livenessCardFrontResult
        
        // Kết quả kiểm tra giấy tờ chụp trực tiếp (Liveness Card) mặt sau
        let livenessCardBackResult = ICEKYCSavedData.shared().livenessCardBackResult
        
        // Dữ liệu thực hiện SO SÁNH khuôn mặt (lấy từ mặt trước ảnh giấy tờ hoặc ảnh thẻ)
        let compareFaceResult = ICEKYCSavedData.shared().compareFaceResult
        
        // Dữ liệu kiểm tra ảnh CHÂN DUNG chụp trực tiếp hay không
        let livenessFaceResult = ICEKYCSavedData.shared().livenessFaceResult
        
        // Dữ liệu kiểm tra ảnh CHÂN DUNG có bị che mặt hay không
        let maskedFaceResult = ICEKYCSavedData.shared().maskedFaceResult
        
        call?.resolve(["ocrResult": ocrResult,
                       "livenessCardFrontResult": livenessCardFrontResult,
                       "livenessCardBackResult": livenessCardBackResult,
                       "livenessFaceResult": livenessFaceResult,
                       "maskedFaceResult": maskedFaceResult,
                       "compareFaceResult": compareFaceResult])
    }
    
    public func icEkycCameraClosed(with type: ScreenType) {
        print("icEkycCameraClosed = \(type)")
    }
    
}

File: capacitor-testapp/android/app/src/main/java/com/capacitorjs/testapp/EkycPlugin.java
package com.capacitorjs.testapp;

import android.app.Activity;
import android.content.Intent;

import androidx.activity.result.ActivityResult;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.vnptit.idg.sdk.activity.VnptIdentityActivity;
import com.vnptit.idg.sdk.utils.KeyIntentConstants;
import com.vnptit.idg.sdk.utils.KeyResultConstants;
import com.vnptit.idg.sdk.utils.SDKEnum;

@CapacitorPlugin(name = "EkycPlugin")
public class EkycPlugin extends Plugin {

   // Phương thức thực hiện eKYC luồng đầy đủ bao gồm: Chụp ảnh giấy tờ và chụp ảnh
   // chân dung
   // Bước 1 - chụp ảnh giấy tờ
   // Bước 2 - chụp ảnh chân dung xa gần
   // Bước 3 - hiển thị kết quả
   @PluginMethod()
   public void startEkycFull(final PluginCall call) {
      final Activity ctx = getActivity();
      if (ctx == null) {
         return;
      }

      final Intent intent = new Intent(ctx, VnptIdentityActivity.class);

      // Nhập thông tin bộ mã truy cập. Lấy tại mục Quản lý Token
      // https://ekyc.vnpt.vn/admin-dashboard/console/project-manager
      intent.putExtra(KeyIntentConstants.ACCESS_TOKEN, call.getString("accessToken", ""));
      intent.putExtra(KeyIntentConstants.TOKEN_ID, call.getString("tokenId", ""));
      intent.putExtra(KeyIntentConstants.TOKEN_KEY, call.getString("tokenKey", ""));

      // Giá trị này dùng để đảm bảo mỗi yêu cầu (request) từ phía khách hàng sẽ
      // không bị thay đổi.
      intent.putExtra(KeyIntentConstants.CHALLENGE_CODE, "INNOVATIONCENTER");

      // Ngôn ngữ sử dụng trong SDK
      // - VIETNAMESE: Tiếng Việt
      // - ENGLISH: Tiếng Anh
      intent.putExtra(KeyIntentConstants.LANGUAGE_SDK, SDKEnum.LanguageEnum.VIETNAMESE.getValue());

      // Bật/Tắt Hiển thị màn hình hướng dẫn
      intent.putExtra(KeyIntentConstants.IS_SHOW_TUTORIAL, true);

      // Bật chức năng hiển thị nút bấm "Bỏ qua hướng dẫn" tại các màn hình hướng dẫn
      // bằng video
      intent.putExtra(KeyIntentConstants.IS_ENABLE_GOT_IT, true);

      // Giá trị này xác định kiểu giấy tờ để sử dụng:
      // - IDENTITY_CARD: Chứng minh thư nhân dân, Căn cước công dân
      // - IDCardChipBased: Căn cước công dân gắn Chip
      // - Passport: Hộ chiếu
      // - DriverLicense: Bằng lái xe
      // - MilitaryIdCard: Chứng minh thư quân đội
      intent.putExtra(KeyIntentConstants.DOCUMENT_TYPE, SDKEnum.DocumentTypeEnum.IDENTITY_CARD.getValue());

      // Bật/Tắt chức năng So sánh ảnh trong thẻ và ảnh chân dung
      intent.putExtra(KeyIntentConstants.IS_COMPARE_FLOW, true);

      // Bật/Tắt chức năng kiểm tra ảnh giấy tờ chụp trực tiếp (liveness card)
      intent.putExtra(KeyIntentConstants.IS_CHECK_LIVENESS_CARD, true);

      // Lựa chọn chức năng kiểm tra ảnh chân dung chụp trực tiếp (liveness face)
      // - NoneCheckFace: Không thực hiện kiểm tra ảnh chân dung chụp trực tiếp hay
      // không
      // - iBETA: Kiểm tra ảnh chân dung chụp trực tiếp hay không iBeta (phiên bản
      // hiện tại)
      // - Standard: Kiểm tra ảnh chân dung chụp trực tiếp hay không Standard (phiên
      // bản mới)
      intent.putExtra(KeyIntentConstants.CHECK_LIVENESS_FACE, SDKEnum.ModeCheckLiveNessFace.iBETA.getValue());

      // Bật/Tắt chức năng kiểm tra che mặt
      intent.putExtra(KeyIntentConstants.IS_CHECK_MASKED_FACE, true);

      // Lựa chọn chế độ kiểm tra ảnh giấy tờ ngay từ SDK
      // - None: Không thực hiện kiểm tra ảnh khi chụp ảnh giấy tờ
      // - Basic: Kiểm tra sau khi chụp ảnh
      // - MediumFlip: Kiểm tra ảnh hợp lệ trước khi chụp (lật giấy tờ thành công →
      // hiển thị nút chụp)
      // - Advance: Kiểm tra ảnh hợp lệ trước khi chụp (hiển thị nút chụp)
      intent.putExtra(KeyIntentConstants.TYPE_VALIDATE_DOCUMENT, SDKEnum.TypeValidateDocument.Basic.getValue());

      // Giá trị này xác định việc có xác thực số ID với mã tỉnh thành, quận huyện,
      // xã phường tương ứng hay không.
      intent.putExtra(KeyIntentConstants.IS_VALIDATE_POSTCODE, true);

      intent.putExtra(KeyIntentConstants.VERSION_SDK, SDKEnum.VersionSDKEnum.ADVANCED.getValue());

      startActivityForResult(call, intent, "ekycFullResult");
   }

   @ActivityCallback
   private void ekycFullResult(final PluginCall call, final ActivityResult result) {
      if (call == null)
         return;
      final Intent data = result.getData();
      if (data == null)
         return;

      /*
       * Dữ liệu bóc tách thông tin OCR
       * {@link KeyResultConstants#INFO_RESULT}
       */
      final String dataInfoResult = data.getStringExtra(KeyResultConstants.INFO_RESULT);

      /*
       * Dữ liệu bóc tách thông tin Liveness card mặt trớc
       * {@link KeyResultConstants#LIVENESS_CARD_FRONT_RESULT}
       */
      final String dataLivenessCardFrontResult = data.getStringExtra(KeyResultConstants.LIVENESS_CARD_FRONT_RESULT);

      /*
       * Dữ liệu bóc tách thông tin liveness card mặt sau
       * {@link KeyResultConstants#LIVENESS_CARD_REAR_RESULT}
       */
      final String dataLivenessCardRearResult = data.getStringExtra(KeyResultConstants.LIVENESS_CARD_REAR_RESULT);

      /*
       * Dữ liệu bóc tách thông tin compare face
       * {@link KeyResultConstants#COMPARE_RESULT}
       */
      final String dataCompareResult = data.getStringExtra(KeyResultConstants.COMPARE_RESULT);

      /*
       * Dữ liệu bóc tách thông tin liveness face
       * {@link KeyResultConstants#LIVENESS_FACE_RESULT}
       */
      final String dataLivenessFaceResult = data.getStringExtra(KeyResultConstants.LIVENESS_FACE_RESULT);

      /*
       * Dữ liệu bóc tách thông tin mask face
       * {@link KeyResultConstants#MASKED_FACE_RESULT}
       */
      final String dataMaskedFaceResult = data.getStringExtra(KeyResultConstants.MASKED_FACE_RESULT);

      final JSObject obj = new JSObject();
      obj.put("ocrResult", dataInfoResult);
      obj.put("livenessCardFrontResult", dataLivenessCardFrontResult);
      obj.put("livenessCardBackResult", dataLivenessCardRearResult);
      obj.put("livenessFaceResult", dataLivenessFaceResult);
      obj.put("maskedFaceResult", dataMaskedFaceResult);
      obj.put("compareFaceResult", dataCompareResult);

      call.resolve(obj);
   }
}
